src/components/ha-code-editor.ts

```diff
@@ -0,0 +1,129 @@
+// @ts-ignore
+import CodeMirror from "codemirror";
+// @ts-ignore
+import codeMirrorCSS from "codemirror/lib/codemirror.css";
+import { fireEvent } from "../common/dom/fire_event";
+
+declare global {
+  interface HASSDomEvents {
+    "editor-save": undefined;
+  }
+}
+
+export class HaCodeEditor extends HTMLElement {
+  public codemirror?: any;
+  private mode: string;
+  private _autofocus = false;
+  private _rtl = false;
+  private _value: string;
+
+  public constructor(mode: string) {
+    super();
+    CodeMirror.commands.save = (cm: CodeMirror) => {
+      fireEvent(cm.getWrapperElement(), "editor-save");
+    };
+    this.mode = mode;
+    this._value = "";
+    const shadowRoot = this.attachShadow({ mode: "open" });
+    shadowRoot.innerHTML = `
+            <style>
+              ${codeMirrorCSS}
+              .CodeMirror {
+                height: var(--code-mirror-height, auto);
+                direction: var(--code-mirror-direction, ltr);
+              }
+              .CodeMirror-scroll {
+                max-height: var(--code-mirror-max-height, --code-mirror-height);
+              }
+              .CodeMirror-gutters {
+                border-right: 1px solid var(--paper-input-container-color, var(--secondary-text-color));
+                background-color: var(--paper-dialog-background-color, var(--primary-background-color));
+                transition: 0.2s ease border-right;
+              }
+              :host(.error-state) .CodeMirror-gutters {
+                border-color: var(--error-state-color, red);
+              }
+              .CodeMirror-focused .CodeMirror-gutters {
+                border-right: 2px solid var(--paper-input-container-focus-color, var(--primary-color));
+              }
+              .CodeMirror-linenumber {
+                color: var(--paper-dialog-color, var(--primary-text-color));
+              }
+              .rtl .CodeMirror-vscrollbar {
+                right: auto;
+                left: 0px;
+              }
+              .rtl-gutter {
+                width: 20px;
+              }
+            </style>`;
+  }
+
+  set value(value: string) {
+    if (this.codemirror) {
+      if (value !== this.codemirror.getValue()) {
+        this.codemirror.setValue(value);
+      }
+    }
+    this._value = value;
+  }
+
+  get value(): string {
+    return this.codemirror ? this.codemirror.getValue() : "";
+  }
+
+  set rtl(rtl: boolean) {
+    this._rtl = rtl;
+    this.setScrollBarDirection();
+  }
+
+  set autofocus(autofocus: boolean) {
+    this._autofocus = autofocus;
+    if (this.codemirror) {
+      this.codemirror.focus();
```

Comment:
This is not what I expect `autofocus` to do. Docs: https://developer.mozilla.org/en-US/docs/Web/API/HTMLSelectElement/autofocus

If we want to stay close to how HTML works, let's check for `autofocus` inside `connectedCallback` and focus if it's true.